### PR TITLE
systemctl installed default on Debian regardless of init system

### PR DESCRIPTION
**Pull request by deHakkelaar**
Before change on a systemd Raspi:

```
pi@noads:~ $ stat /sbin/init
  File: ‘/sbin/init’ -> ‘/lib/systemd/systemd’
```

```
pi@noads:~ $ pihole restartdns
pi@noads:~ $
```

Before change on an upstart Raspi:

```
xbian@monmc ~ $ /sbin/init --version
init (upstart 1.13.2)
```

```
xbian@monmc ~ $ pihole restartdns
Failed to restart dnsmasq.service: No such method 'RestartUnit'
See system logs and 'systemctl status dnsmasq.service' for details.
xbian@monmc ~ $
```

After change on a systemd Raspi:

```
pi@noads:~ $ pihole restartdns
pi@noads:~ $
```

After change on an upstart Raspi:

```
xbian@monmc ~ $ pihole restartdns
[ ok ] Restarting DNS forwarder and DHCP server: dnsmasq.
xbian@monmc ~ $
```

**By submitting this pull request, I confirm the following (please check boxes, eg [X]) _Failure to fill the template will close your PR_:**

***Please submit all pull requests against the `development` branch. Failure to do so will delay or deny your request***

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/pi-hole/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:**

_6_

---
_systemctl installed default on Debian regardless of init system_

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

### CODE REVIEW

**Comment by dschaper** [comment]
This should solve the XBian issue, correct?

**Comment by deHakkelaar** [comment]
Correct.
XBian uses upstart for init.

**Comment by mkreisl** [comment]
> This should solve the XBian issue, correct?

Not only XBian issue, required if init system is **not** systemd

**Comment by dschaper** [comment]
We had a check in the code for non-systemd based installs, it just got modified to check for the systemd binaries as a check, instead of checking to see what the actual init system was that was in play. We run in to similar problems with Synology chroot/jailed installs where the jailed `systemd` isn't the true init system.

**Comment by dschaper** [comment]
Can you rebase your pull off of the Development branch, we don't allow PR's against master.

**Comment by deHakkelaar** [comment]
The dev branch is altered already in such a way that I am not sure if I can apply the original patch on this dev branch.
The dev branch depends on systemctl (systemd) not showing any output if a dnsmasq restart is executed successfully.
With upstart, there is always an output returned, not only on error, as can be seen from below.

Systemd without and with an error introduced:

```
pi@noads:~ $ pihole restartdns
pi@noads:~ $
```

```
pi@noads:~ $ pihole restartdns
Job for dnsmasq.service failed. See 'systemctl status dnsmasq.service' and 'journalctl -xn' for details.
```

Upstart without and with an error introduced:

```
xbian@monmc ~ $ pihole restartdns
[ ok ] Restarting DNS forwarder and DHCP server: dnsmasq.
xbian@monmc ~ $
```

```
xbian@monmc ~ $ pihole restartdns
[FAIL] Restarting DNS forwarder and DHCP server: configuration syntax check failed!
xbian@monmc ~ $
```

Below bit, that was added to the dev branch, in the restartDNS function will always be FALSE with an upstart or systemv systems:

```
    if [[ -z "${output}" ]]; then
      echo -e "${OVER}  ${TICK} ${str}"
    else
      echo -e "${OVER}  ${CROSS} ${output}"
    fi
```

Not sure what to do except install the dev branch on two Pi's for testing.

**Comment by WaLLy3K** [comment]
So the issue here is that `systemctl` is useless, and we want to get rid of it? The following code will read the `$?` exit code of `service`, and only print the output when the exit code is not 0:

(I also took a moment to clean up the restart service code)
```
restartDNS() {
  local dnsmasqPid
  dnsmasqPid=$(pidof dnsmasq)

  # Get PID of dnsmasq to determine if it needs to be started or restarted
  if [[ "${dnsmasqPid}" ]]; then
    svcOption="restart"
  else
    svcOption="start"
  fi

  local str="${svcOption^}ing DNS service"
  echo -ne "  ${INFO} ${str}..."

  if [[ -x "$(command -v systemctl)" ]]; then
    output=$( { systemctl "${svcOption}" dnsmasq; } 2>&1 )
    status="$?"
  else
    output=$( { service dnsmasq "${svcOption}"; } 2>&1 )
    status="$?"
  fi

  if [[ "${status}" -eq 0 ]]; then
    echo -e "${OVER}  ${TICK} ${str}"
  else
    echo -e "${OVER}  ${CROSS} ${output}"
  fi
}
```

Would this achieve what you're looking for, or is there something else I'm missing? :smile:

**Comment by deHakkelaar** [comment]
Looking way better.
I dropped that code snippet on both of my stable Pi-hole systems (one is upstart and other is systemd).

Systemd:

```
pi@noads:~ $ pihole restartdns
   Restarting DNS service...   Restarting DNS service
```

Systemd with a forced error:

```
pi@noads:~ $ pihole restartdns
   Restarting DNS service...   Job for dnsmasq.service failed. See 'systemctl status dnsmasq.service' and 'journalctl -xn' for details.
```

Upstart:

```
xbian@monmc ~ $ pihole restartdns
   Restarting DNS service...   Restarting DNS service
xbian@monmc ~ $
```

Upstart whith a forced error:

```
xbian@monmc ~ $ pihole restartdns
   Restarting DNS service...   Restarting DNS forwarder and DHCP server: configuration syntax check failed!
xbian@monmc ~ $
```

But with updating the lists, I still have below bug for upstart when I force an error.

Systemd:

```
pi@noads:~ $ pihole -g
::: Refresh lists in dnsmasq...   Starting DNS service...   Starting DNS service
 done!
::: DNS service is running
::: Pi-hole blocking is Enabled
pi@noads:~ $
```

Systemd with a forced error:

```
pi@noads:~ $ pihole -g
::: Refresh lists in dnsmasq...   Restarting DNS service...   Job for dnsmasq.service failed. See 'systemctl status dnsmasq.service' and 'journalctl -xn' for details.
 done!
::: DNS service is NOT running
pi@noads:~ $
```

Upstart:

```
xbian@monmc ~ $ pihole -g
::: Refresh lists in dnsmasq...   Restarting DNS service...   Restarting DNS service
 done!
::: DNS service is running
::: Pi-hole blocking is Enabled
```

Upstart with a forced error:

```
xbian@monmc ~ $ pihole -g
::: Refresh lists in dnsmasq...   Restarting DNS service...   Restarting DNS forwarder and DHCP server: configuration syntax check failed!
 done!
::: DNS service is running
::: Pi-hole blocking is Enabled
```

I'll have a look if I can figure this out.

**Comment by WaLLy3K** [comment]
> pi@noads:~ $ pihole restartdns
   Restarting DNS service...   Restarting DNS service

Does your system show a new line between `...   Restarting` that failed to paste across to GitHub?

Without those newlines, it makes it mighty hard to follow! :smile:

**Comment by deHakkelaar** [comment]
No newline is displayed.
The output on upstart is exactly the same as on systemd (if no errors).

**Comment by WaLLy3K** [comment]
If you could provide a screenshot of the issue you're facing, I should be able to better understand what's happening!

**Comment by deHakkelaar** [comment]
The issue with the sippet is that on upstart, pihole believes dnsmasq is still running when I force a dnsmasq error (syntax check failed):

```
xbian@monmc ~ $ pihole -g
::: Refresh lists in dnsmasq...   Restarting DNS service...   Restarting DNS forwarder and DHCP server: configuration syntax check failed!
 done!
::: DNS service is running
::: Pi-hole blocking is Enabled
```

Do you mean you want to see a screenshot of a successful dnsrestart on upstart as well as systemd ?

**Comment by WaLLy3K** [comment]
> The issue with the sippet is that on upstart, pihole believes dnsmasq is still running

If there's a result for `pidof dnsmasq`, then it's going to believe it's up and running before it runs any command.

As far as screenshots go, I just want to see any errors that are cropping up because I'm still failing to comprehend what's actually happening here (Though 3AM will do that to you!)

**Comment by deHakkelaar** [comment]
My mistake.
dnsmasq keeps on running if I force an error by mangling the dnsmasq.conf file and restart dnsmasq (pidof dnsmasq).
I was expecting dnsmasq refusing to start/restart if I force an error but it does not.
I'll investigate/test a bit more and let you know.

**Comment by jacobsalmela** [comment]
@deHakkelaar did @WaLLy3K's PR #1630 fix this for you?  Please let me know since this PR has gone stale.

**Comment by deHakkelaar** [comment]
Sorry, I fell ill and forgot about it.
I'll rig a sandbox to check and let you know.

**Comment by deHakkelaar** [comment]
Preliminary findings not good.
Previously, I just had to do this to get Pi-hole running on upstart:

https://discourse.pi-hole.net/t/ftl-on-xbian-pi/2882/15?u=dehakkelaar

Now I have issues with dnsmasq and lighttpd just installing stable:

http://paste.ubuntu.com/26065895/

And luckily dhcpcd5 didnt activate or it would probably have broken the XBian configuration tools and Kodi XBian add-on that allows to set network settings.
Will investigate bit more.

Install log here:

http://paste.ubuntu.com/26065912/

**Comment by deHakkelaar** [comment]
Oops, lighttpd just needed a reboot.
Other issues remain.
